apache/spark
# [SPARK-38049][K8S][TESTS] Use Java 17 in K8s integration tests by default

### What changes were proposed in this pull request?

This PR aims to use `Java 17` in K8s integration tests by default.

### Why are the changes needed?

Java 8 cannot run Java11/17-built Spark distribution.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?

Manually run the following and check the Java version in the generated docker images.

**SBT**
```
$ build/sbt -Psparkr -Pkubernetes -Pkubernetes-integration-tests -Dtest.exclude.tags=minikube -Dspark.kubernetes.test.deployMode=docker-for-desktop "kubernetes-integration-tests/test"
...
[info] KubernetesSuite:
[info] - Run SparkPi with no resources (8 seconds, 949 milliseconds)
[info] - Run SparkPi with no resources & statefulset allocation (8 seconds, 515 milliseconds)
[info] - Run SparkPi with a very long application name. (8 seconds, 389 milliseconds)
[info] - Use SparkLauncher.NO_RESOURCE (8 seconds, 393 milliseconds)
[info] - Run SparkPi with a master URL without a scheme. (8 seconds, 360 milliseconds)
[info] - Run SparkPi with an argument. (8 seconds, 435 milliseconds)
[info] - Run SparkPi with custom labels, annotations, and environment variables. (8 seconds, 611 milliseconds)
[info] - All pods have the same service account by default (8 seconds, 353 milliseconds)
[info] - Run extraJVMOptions check on driver (4 seconds, 364 milliseconds)
[info] - Run SparkRemoteFileTest using a remote data file (8 seconds, 392 milliseconds)
[info] - Verify logging configuration is picked from the provided SPARK_CONF_DIR/log4j2.properties (14 seconds, 564 milliseconds)
[info] - Run SparkPi with env and mount secrets. (16 seconds, 868 milliseconds)
[info] - Run PySpark on simple pi.py example (9 seconds, 632 milliseconds)
[info] - Run PySpark to test a pyfiles example (10 seconds, 520 milliseconds)
[info] - Run PySpark with memory customization (8 seconds, 385 milliseconds)
[info] - Run in client mode. (7 seconds, 336 milliseconds)
[info] - Start pod creation from template (8 seconds, 727 milliseconds)
[info] - Test basic decommissioning (42 seconds, 353 milliseconds)
[info] - Test basic decommissioning with shuffle cleanup (42 seconds, 532 milliseconds)
[info] - Test decommissioning with dynamic allocation & shuffle cleanups (2 minutes, 40 seconds)
[info] - Test decommissioning timeouts (42 seconds, 211 milliseconds)
[info] - SPARK-37576: Rolling decommissioning (1 minute, 7 seconds)
[info] - Run SparkR on simple dataframe.R example (12 seconds, 16 milliseconds)
[info] Run completed in 11 minutes, 24 seconds.
[info] Total number of tests run: 23
[info] Suites: completed 1, aborted 0
[info] Tests: succeeded 23, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
```

**MAVEN**
```
$ mvn package -Pkubernetes -DskipTests
$ resource-managers/kubernetes/integration-tests/dev/dev-run-integration-tests.sh --deploy-mode docker-for-desktop --namespace default --exclude-tags minikube,r
...
KubernetesSuite:
- Run SparkPi with no resources
- Run SparkPi with no resources & statefulset allocation
- Run SparkPi with a very long application name.
- Use SparkLauncher.NO_RESOURCE
- Run SparkPi with a master URL without a scheme.
- Run SparkPi with an argument.
- Run SparkPi with custom labels, annotations, and environment variables.
- All pods have the same service account by default
- Run extraJVMOptions check on driver
- Run SparkRemoteFileTest using a remote data file
- Verify logging configuration is picked from the provided SPARK_CONF_DIR/log4j2.properties
- Run SparkPi with env and mount secrets.
- Run PySpark on simple pi.py example
- Run PySpark to test a pyfiles example
- Run PySpark with memory customization
- Run in client mode.
- Start pod creation from template
- Test basic decommissioning
- Test basic decommissioning with shuffle cleanup
- Test decommissioning with dynamic allocation & shuffle cleanups
- Test decommissioning timeouts
- SPARK-37576: Rolling decommissioning
Run completed in 8 minutes, 52 seconds.
Total number of tests run: 22
Suites: completed 2, aborted 0
Tests: succeeded 22, failed 0, canceled 0, ignored 0, pending 0
All tests passed.
```

```
$ docker run -it --rm kubespark/spark:3.3.0-SNAPSHOT_C0A4AD5A-2561-4972-B2DE-0FDA941F8064 java -version | tail -n1
OpenJDK 64-Bit Server VM (build 17.0.2+8-Debian-1deb11u1, mixed mode, sharing)
```
